### PR TITLE
Setup TypeScript Project References

### PR DESCRIPTION
**Pull request by mohsen1**
This change introduces [Project Reference](https://www.typescriptlang.org/docs/handbook/project-references.html) makes it possible to type-check each package against other packages. In future we can use the `incremental` compiler option to speed up compilations. 

It also allows editors like VSCode understand that packages like `"@redwood/internal"` are local so you can click through identifiers and import specifiers and go to the package source. 

I've removed `"build:types"` tasks because those are not necessary anymore. The top level TS build will populate the types. In development, the TS server will emit declaration files as well. 

This paves the way for more TypeScript conversion. I've converted the `core` package to TypeScript in this PR. Other packages deserve to have their own PR after we iron out all of the details here. 

Kapture 2020-03-29 at 18 17 17

### CODE REVIEW

**Comment by peterp** [comment]
This is awesome, thank you!

I need a bit of time to catch up here, but I'll review this shortly!

**Comment by thedavidprice** [comment]
Our CI Checks were not running for PRs from external, forked repos. See #358 

This is now fixed but will not retroactively apply to this PR. Before merging, please locally confirm the following checks pass:
- [ ] `yarn build`
- [ ] `yarn lint`
- [ ] `yarn test`

Thanks in advance!

**Comment by mohsen1** [comment]
CI is running now thanks for fixing that. This must merge after https://github.com/redwoodjs/redwood/pull/363 is merged.

**Comment by thedavidprice** [comment]
> This must merge after #363 is merged.

@mohsen1 Thanks for seeing that! I noticed yesterday and was aware there’d be conflicts, which I’ll handle if PRs happen out of order. Adding dependency label. And will try to get a test done on Windows soon 🤔

File: packages/dev-server/src/main.ts [hunk 1]
```diff
@@ -183,7 +183,7 @@ app.all('/:routeName', async (req, res) => {
   }
 })
 
-const reloadLambdas = (path: string) => {
+const reloadLambdas = (_path: string) => {
```
Comment (peterp) [inline packages/dev-server/src/main.ts:186]: I don't seem to be using that at all.
```suggestion
const reloadLambdas = () => {
```

File: packages/core/src/loaders/routes-auto-loader.ts [hunk 2]
```diff
@@ -1,26 +1,27 @@
...
   // right thing with watched files, etc.
-  filteredDeps.forEach((entry) => {
+  filteredDeps.forEach(entry => {
     this.addDependency(entry.path)
   })
 
   // Grab the import strings from the deps and get them ready to be appended
   // onto the Routes file.
-  const importString = filteredDeps.map((x) => x.importStatement).join('\n')
+  const importString = filteredDeps.map(x => x.importStatement).join('\n')
```
[the remark this answers is not in the document]
Comment (mohsen1) [reply packages/core/src/loaders/routes-auto-loader.ts:24]: I've added `yarn prettier --write .` to the `"fix"` script. Running it updated so many files!

[the remark this answers is not in the document]
Comment (peterp) [reply packages/core/src/loaders/routes-auto-loader.ts]: Ah man, I'm sorry, I should've been clearer. We run Prettier via ESLint.

In order to run ESLint you have to build the packages that the config relies on:


```terminal
cd redwood.git
yarn build
yarn lint:fix // this should work
```

[the remark this answers is not in the document]
Comment (mohsen1) [reply packages/core/src/loaders/routes-auto-loader.ts]: Gotcha! I'll update the scripts once the Windows PR is merged.

**Comment by thedavidprice** [comment]
#363 has been merged. There are the expects conflicts to now resolve with the `package.json` scripts, after which this is ready to merge. Thanks again!

**Comment by mohsen1** [comment]
Running Prettier with ESLint is great but it doesn't cover other file types such as YAML and Markdown. How about I add prettier executable in npm scripts for `lint` and `lint:fix`?

**Comment by thedavidprice** [comment]
@mohsen1 Re: "adding prettier executable" I'm looping in @peterp 

In general, it makes sense to add coverage for all the file types we use. And I hadn't thought about config or markdown, so this is a good catch. But I haven't previously used Prettier alone or attempted to add executable alongside ESlint. Some questions to that end:
- Instead of adding the Prettier executable, is it possible to add file types to our ESLint config?
- If ESlint and Prettier are run in series, are there issues to be aware of with them possibly "working against each other?", e.g. how do we ensure they don't overlap or duplicate checks?
- If we include file types to check during `yarn lint`, can we ensure those files are also being auto-checked during editing and save? It would be a less than ideal situation to only experience errors when the linter is run as a script or during CI.

Thanks again for your work here!

**Comment by mohsen1** [comment]
I did not understand prettier-eslint properly actually. [According to their docs](https://github.com/prettier/prettier-eslint) it can format non-js/ts files too:

> For files with an extension of .css, .less, .scss, or .json this only runs prettier since eslint cannot process those.

I'll try removing direct dependency to prettier and calling prettier in npm scripts now.

**Comment by mohsen1** [comment]
It was quick to run a corss-platform [issues with prettier-eslint](https://github.com/prettier/prettier-eslint-cli/issues/208)! 

> If ESlint and Prettier are run in series, are there issues to be aware of with them possibly "working against each other?", e.g. how do we ensure they don't overlap or duplicate checks?

This should not be an issue because the eslint config has the prettier plugin on top

> If we include file types to check during yarn lint, can we ensure those files are also being auto-checked during editing and save? It would be a less than ideal situation to only experience errors when the linter is run as a script or during CI.

Prettier editor integration is using the same engine and configuration as this CLI. This also should not be an issue. We have the same setup at my workplace. 

I'm keeping it as-is because this works cross-platform.

File: packages/web/src/form/README.md [hunk 3]
```diff
@@ -84,7 +84,7 @@ const ContactPage = () => {
 Any form you want Redwood to provide validation and error styling on should be surrounded by this tag. Except for the view attributes specific to validation and submission, props are passed down to the regular `<form>` tag that is rendered.
 
 ```html
-<Form onSubmit={onSubmit} className="form">...</Form>
+<form onSubmit="{onSubmit}" className="form">...</form>
```
Comment (peterp) [inline packages/web/src/form/README.md:87]: I'm guessing this has to be `jsx` to format correctly?

**Comment by peterp** [comment]
I would prefer not to require a particular style of formatting for our non-src files, we got a lot of people contributing small fixes and I think it would generate a bunch of noise to have their "Update README" pull-request fail because of a formatting issue.

I really like that formatting is uniformly applied to the configuration files, but I do think that these changes are out of scope for this particular pull-request.

**Comment by thedavidprice** [comment]
@mohsen1 Is this one ready for @peterp to re-review? If so, we should be able to get it merged before the next release.

**Comment by mohsen1** [comment]
@thedavidprice yes this is ready.

File: docs/assets.md [hunk 4]
```diff
@@ -44,9 +49,11 @@ Running `yarn build` will copy the file `favicon.png` to `/dist/favicon.png`. Th
 ```html
 <link rel="icon" type="image/png" href="/favicon.png" />
 ```
+
 and
+
 ```html
-<img src="/static-files/my-logo.jpg"> alt="Logo" />
+<img src="/static-files/my-logo.jpg" /> alt="Logo" />
```
Comment (peterp) [inline docs/assets.md:56]:
```suggestion
<img src="/static-files/my-logo.jpg" alt="Logo" />
```

File: package.json [hunk 5]
```diff
@@ -20,6 +20,8 @@
     "jest": "^25.2.3",
     "lerna": "^3.20.2",
     "nodemon": "^2.0.2",
+    "npm-run-all": "^4.1.5",
+    "prettier": "2",
```
Comment (peterp) [inline package.json:24]:
```suggestion
    "prettier": "^2.0.4",
```

File: package.json [hunk 6]
```diff
@@ -30,9 +32,16 @@
     "packages/api/importAll.macro.js"
   ],
   "scripts": {
-    "build": "yarn cross-env NODE_ENV=production lerna run build",
+    "build": "run-s clean build:js build:types",
+    "clean": "lerna run clean",
+    "build:types": "tsc --build --clean && tsc --build",
+    "build:js": "cross-env NODE_ENV=production lerna run build",
```
Comment (peterp) [inline package.json:38]:
```suggestion
    "build:js": "cross-env NODE_ENV=production lerna run build",
    "build:types": "tsc --build --clean && tsc --build",
```

File: package.json [hunk 7]
```diff
@@ -30,9 +32,16 @@
...
     "test": "lerna run test --stream -- --colors",
-    "lint": "yarn eslint packages",
-    "lint:fix": "yarn eslint --fix packages"
+    "lint": "run-s eslint:check prettier:check",
+    "lint:fix": "run-s eslint:fix prettier:fix",
+    "eslint:check": "yarn eslint packages",
+    "eslint:fix": "yarn eslint --fix packages",
+    "prettier:check": "prettier --list-different .",
+    "prettier:fix": "prettier --write ."
```
Comment (peterp) [inline package.json:45]:
```suggestion
```

File: package.json [hunk 8]
```diff
@@ -30,9 +32,16 @@
     "packages/api/importAll.macro.js"
   ],
   "scripts": {
-    "build": "yarn cross-env NODE_ENV=production lerna run build",
+    "build": "run-s clean build:js build:types",
+    "clean": "lerna run clean",
```
Comment (peterp) [inline package.json:36]:
```suggestion
    "build:clean": "lerna run build:clean",
```

File: package.json [hunk 9]
```diff
@@ -30,9 +32,16 @@
     "packages/api/importAll.macro.js"
   ],
   "scripts": {
-    "build": "yarn cross-env NODE_ENV=production lerna run build",
+    "build": "run-s clean build:js build:types",
```
Comment (peterp) [inline package.json:35]:
```suggestion
    "build": "run-s build:clean build:js build:types",
```

File: package.json (hunk 7 above)
Comment (peterp) [inline package.json:45]:
```suggestion
    "lint": "yarn eslint packages",
    "lint:fix": "yarn eslint --fix packages"
```

**Comment by peterp** [comment]
@mohsen1 Rad, `&&` actually works on Windows. I'm inclined to remove `npm-run-all` since it does so little here.

**Comment by mohsen1** [comment]
I didn't realize `&&` works. Apparently `&` doesn't work but `&&` does. TIL!

**Comment by peterp** [comment]
I was suprised myself, semi-colons are the things that don't work in Windows commands.

**Comment by peterp** [comment]
I'll resolve these conflicts and get this merged in the morning.

**Pull request merged**